File: extra/r/excon.r
# EXCON interpreter (R cross-check; see README "Extra Implementations").
#
# An 8-cell bit pool with a pointer: ":" resets the pool and pointer, "^"
# flips the current bit, "<" moves the pointer down, and "!" prints the pool
# as a binary byte (MSB first).  The language is straight-line with no
# control flow.
#
# Deviations: the pool is indexed 1..8 with the pointer starting at 8 and "<"
# decrementing it with no bounds check, so a program with more than 7 "<"s
# runs off the pool rather than faulting (the Python interpreter raises
# HaltError instead).  R's stdout cannot carry a NUL byte, so "!" of an
# all-zero pool prints nothing rather than the NUL the Python interpreter
# writes.  "cat" is used instead of a single output channel.
#
# Input: the program file is the first command-line argument; the language
# has no input command.

file <- commandArgs(trailingOnly = TRUE)[1]
size <- file.info(file)$size
strg <- readChar(file, size)
syms <- strsplit(strg, "")[[1]]


tape <- rep(c(0), times = 8)
cell <- 8


bin <- function(arr) {
  sum <- 0
  for (k in c(1:8))
  if (tape[k])
    sum <- sum + 2 ^ (8 - k)
  sum
}


for (char in syms) {
  if (char == ":") {
    cell <- 8
    tape <- rep(c(0), times = 8)
  } else if (char == "^") {
    tape[cell] <- 1 - tape[cell]
  } else if (char == "!") {
    num <- bin(tape)
    cat(intToUtf8(num))
  } else if (char == "<") {
    cell <- cell - 1
  }
}

File: src/esolangs/interpreters/tape_based/excon.py
"""Interpreter for EXCON.

An 8-cell bit pool with a pointer: : resets the pool and pointer, ^ flips the
current bit, < decrements the pointer, and ! prints the pool as a binary byte
(MSB first).  The language is straight-line with no control flow.

The wiki calls moving the pointer more than 8 steps left a fault; this
interpreter treats it as an invalid operation and halts the program with
:class:`~esolangs.exceptions.HaltError`.
"""

import sys

from esolangs.exceptions import HaltError
from esolangs.interpreters.io import IO


def run(code: str, io: IO) -> None:
    """Run an EXCON program, printing each pool as a byte."""
    pool: list[int] = [0] * 8
    cell = 7

    for sym in code:
        if sym == ":":
            pool, cell = ([0] * 8, 7)
        elif sym == "^":
            pool[cell] ^= 1
        elif sym == "!":
            num = "".join(map(str, pool))
            io.print_char(chr(int(num, 2)))
        elif sym == "<":
            cell -= 1
            if cell < 0:
                raise HaltError


if __name__ == "__main__":
    if len(sys.argv) > 1:
        with open(sys.argv[1]) as file:
            data = file.read()
            run(data, IO())
